### PR TITLE
Get latestRun from queue on updateProducer and prnUpdate function

### PR DESCRIPTION
**Pull request by Ehsan-Hatami**

### CODE REVIEW

File: src/EprPrnIntegration.Common/Service/ServiceBusProvider.cs [hunk 1]
```diff
@@ -1,45 +1,126 @@
...
-                await using var sender = _serviceBusClient.CreateSender(_serviceBusConfig.Value.FetchPrnQueueName);
+                await using var sender = serviceBusClient.CreateSender(config.Value.FetchPrnQueueName);
                 using ServiceBusMessageBatch messageBatch = await sender.CreateMessageBatchAsync();
                 foreach (var prn in prns)
                 {
                     var jsonPrn = JsonSerializer.Serialize(prn);
                     if (!messageBatch.TryAddMessage(new ServiceBusMessage(jsonPrn)))
                     {
-                        _logger.LogWarning("SendFetchedNpwdPrnsToQueue - The message with EvidenNo: {EvidenNo} is too large to fit in the batch.", prn.EvidenceNo);
+                        logger.LogWarning("SendFetchedNpwdPrnsToQueue - The message with EvidenNo: {EvidenNo} is too large to fit in the batch.", prn.EvidenceNo);
```
Comment (stephan-williamson) [inline src/EprPrnIntegration.Common/Service/ServiceBusProvider.cs:27]: I don't know the usage so its not clear to me whether {EvidenNo} is correct or it should be {EvidenceNo}. If it should be {EvidenceNo} please change

File: src/EprPrnIntegration.Api/UpdateProducersFunction.cs [hunk 2]
```diff
@@ -2,27 +2,33 @@
...
-        int startHour = GetStartHour(configuration["UpdateProducersStartHour"]);
-
-        var toDate = DateTime.Today.AddHours(startHour);
-        var fromDate = toDate.AddDays(-1);
+        var deltaRun = await DeltaSyncExecution();
 
+        var toDate = DateTime.UtcNow;
+        var fromDate = deltaRun.LastSyncDateTime;
+         
         logger.LogInformation($"Fetching producers from {fromDate} to {toDate}.");
```
Comment (stephan-williamson) [inline src/EprPrnIntegration.Api/UpdateProducersFunction.cs:32]: String interpolation $ is not as efficient in loggers, better log("{X}", x) than log($({x}");

File: src/EprPrnIntegration.Common/Service/ServiceBusProvider.cs [hunk 3]
```diff
@@ -1,45 +1,126 @@
...
             }
             catch (Exception ex)
             {
-                _logger.LogError("SendFetchedNpwdPrnsToQueue failed to add message on Queue with exception: {exception}", ex);
+                logger.LogError("SendDeltaSyncExecutionToQueue failed to add message on Queue with exception: {exception}", ex);
                 throw;
             }
         }
+
+        public async Task<DeltaSyncExecution?> ReceiveDeltaSyncExecutionFromQueue(NpwdDeltaSyncType syncType)
```
Comment (stephan-williamson) [inline src/EprPrnIntegration.Common/Service/ServiceBusProvider.cs:82]: A comment would help readers who don't necessarily know the purpose of "ReceiveDeltaSyncExecutionFromQueue" which is a bit of a mouth full

File: src/EprPrnIntegration.Api.UnitTests/UpdateProducersFunctionTests.cs [hunk 4]
```diff
@@ -76,10 +75,20 @@ public async Task Run_NoUpdatedProducers_LogsWarning()
         _configurationMock.Setup(c => c["UpdateProducersStartHour"]).Returns(startHour.ToString());
```
Comment (Ravi-SharmaGautam) [inline src/EprPrnIntegration.Api.UnitTests/UpdateProducersFunctionTests.cs:75]: Can we remove this setup as we are not using this config anymore

File: src/EprPrnIntegration.Api.UnitTests/UpdateProducersFunctionTests.cs [hunk 5]
```diff
@@ -151,8 +183,54 @@ public async Task Run_FailedToUpdateProducers_LogsRawResponseBody()
...
+            _configurationMock.Object,
+            _serviceBusProviderMock.Object);
+
+        // Act
+        await function.Run(null);
+
+        // Assert
+        _serviceBusProviderMock.Verify(provider => provider.SendDeltaSyncExecutionToQueue(It.Is<DeltaSyncExecution>(d => d.SyncType == NpwdDeltaSyncType.UpdatedProducers)), Times.Once);
+    }
 }
```
Comment (Ravi-SharmaGautam) [inline src/EprPrnIntegration.Api.UnitTests/UpdateProducersFunctionTests.cs:236]: Can we add test to validate if there is no message on queue it is picking default from config

File: src/EprPrnIntegration.Api/UpdateProducersFunction.cs [hunk 6]
```diff
@@ -65,7 +76,8 @@ private async Task HandleApiResponse(HttpResponseMessage pEprApiResponse, DateTi
     {
         if (pEprApiResponse.IsSuccessStatusCode)
         {
-            logger.LogInformation($"Producers list successfully updated in NPWD for time period {fromDate} to {toDate}.");
+            logger.LogInformation(
+                $"Producers list successfully updated in NPWD for time period {fromDate} to {toDate}.");
             return;
         }
```
Comment (Ravi-SharmaGautam) [inline src/EprPrnIntegration.Api/UpdateProducersFunction.cs:82]: I think as part of this ticket similar changes need to be done on updateprns can we do that as well?

File: src/EprPrnIntegration.Common/Service/ServiceBusProvider.cs [hunk 7]
```diff
@@ -1,45 +1,126 @@
...
+                throw;
+            }
+        }
+
+        public async Task SendDeltaSyncExecutionToQueue(DeltaSyncExecution deltaSyncExecution)
+        {
+            try
+            {
+                await using var sender = serviceBusClient.CreateSender(config.Value.DeltaSyncQueueName);
+                using var messageBatch = await sender.CreateMessageBatchAsync();
```
Comment (Ravi-SharmaGautam) [inline src/EprPrnIntegration.Common/Service/ServiceBusProvider.cs:45]: I don't think we are going to have batch here we are pushing only one message so no need of batch which will save additional logic around try add

File: src/EprPrnIntegration.Api.UnitTests/UpdateProducersFunctionTests.cs [hunk 8]
```diff
@@ -16,26 +18,9 @@ public class UpdateProducersFunctionTests
...
-            LogLevel.Error,
-            It.IsAny<EventId>(),
-            It.Is<It.IsAnyType>((v, t) => v.ToString().Contains("Invalid StartHour configuration value")),
-            null,
-            (Func<It.IsAnyType, Exception, string>)It.IsAny<object>()), Times.Once);
-    }
-
+    // Test: Check if the function handles valid StartHour properly
     [Fact]
     public async Task Run_ValidStartHour_FetchesAndUpdatesProducers()
```
[the remark this answers is not in the document]
Comment (Ehsan-Hatami) [reply src/EprPrnIntegration.Api.UnitTests/UpdateProducersFunctionTests.cs:25]: The test is not valid anymore and its deleted

File: src/EprPrnIntegration.Api.UnitTests/UpdateProducersFunctionTests.cs (hunk 4 above)
Comment (Ehsan-Hatami) [reply src/EprPrnIntegration.Api.UnitTests/UpdateProducersFunctionTests.cs:75]: done

File: src/EprPrnIntegration.Api.UnitTests/UpdateProducersFunctionTests.cs (hunk 5 above)
Comment (Ehsan-Hatami) [reply src/EprPrnIntegration.Api.UnitTests/UpdateProducersFunctionTests.cs:236]: done

File: src/EprPrnIntegration.Api/UpdateProducersFunction.cs [hunk 9]
```diff
@@ -36,16 +42,20 @@ public async Task Run([HttpTrigger(AuthorizationLevel.Anonymous, "get")] HttpReq
         var pEprApiResponse = await npwdClient.Patch(npwdUpdatedProducers, NpwdApiPath.UpdateProducers);
 
         await HandleApiResponse(pEprApiResponse, fromDate, toDate);
+
+        deltaRun.LastSyncDateTime = DateTime.UtcNow;
```
[the remark this answers is not in the document]
Comment (Ehsan-Hatami) [reply src/EprPrnIntegration.Api/UpdateProducersFunction.cs:46]: done

File: src/EprPrnIntegration.Api/UpdateProducersFunction.cs (hunk 2 above)
Comment (Ehsan-Hatami) [reply src/EprPrnIntegration.Api/UpdateProducersFunction.cs:32]: done

File: src/EprPrnIntegration.Api/UpdateProducersFunction.cs (hunk 6 above)
Comment (Ehsan-Hatami) [reply src/EprPrnIntegration.Api/UpdateProducersFunction.cs:82]: done

File: src/EprPrnIntegration.Common/Service/ServiceBusProvider.cs (hunk 1 above)
Comment (Ehsan-Hatami) [reply src/EprPrnIntegration.Common/Service/ServiceBusProvider.cs:27]: I reckon this is referring to Eviden as the company number. not sure tbh as its an existing code and I only changed the logger and serviceBusClient variable name

Comment (Ravi-SharmaGautam) [reply src/EprPrnIntegration.Common/Service/ServiceBusProvider.cs:27]: We can ignore this it should be evidenceNo but this is going to change with upcoming ticket which venkata is working

File: src/EprPrnIntegration.Common/Service/ServiceBusProvider.cs (hunk 3 above)
Comment (Ehsan-Hatami) [reply src/EprPrnIntegration.Common/Service/ServiceBusProvider.cs:74]: added comments

**Review by Ravi-SharmaGautam — APPROVED** [review]
LGTM

**Pull request merged**